File: e2e/testdata/lifecycle/python/hooks.py
import threading
import time


def new():
    return Function()


class Function:
    def __init__(self):
        self.start_called = False
        self.config_value = ""
        self._ready = False
        self._alive = True
        self._nonce = str(time.time_ns())
        threading.Timer(15.0, self._become_ready).start()

    def _become_ready(self):
        self._ready = True

    def start(self, cfg):
        self.start_called = True
        self.config_value = cfg.get("TEST_CONFIG_VALUE", "")

    def ready(self):
        return self._ready, "ready" if self._ready else "warming up"

    def alive(self):
        return self._alive, "alive" if self._alive else "unhealthy"

    async def handle(self, scope, receive, send):
        path = scope.get("path", "/")

        if path == "/set-unhealthy":
            self._alive = False
            body = b"UNHEALTHY"
        elif self.start_called:
            body = f"START_OK:{self.config_value} READY={str(self._ready).lower()} ALIVE={str(self._alive).lower()} NONCE={self._nonce}".encode()
        else:
            body = f"START_NOT_CALLED READY={str(self._ready).lower()} ALIVE={str(self._alive).lower()} NONCE={self._nonce}".encode()

        await send({
            "type": "http.response.start",
            "status": 200,
            "headers": [[b"content-type", b"text/plain"]],
        })
        await send({
            "type": "http.response.body",
            "body": body,
        })

File: e2e/testdata/lifecycle/go/hooks.go
package function

import (
	"context"
	"fmt"
	"net/http"
	"sync"
	"time"
)

type Function struct {
	mu          sync.Mutex
	startCalled bool
	configValue string
	ready       bool
	alive       bool
	nonce       string
}

func New() *Function {
	f := &Function{
		alive: true,
		nonce: fmt.Sprintf("%d", time.Now().UnixNano()),
	}
	go func() {
		time.Sleep(15 * time.Second)
		f.mu.Lock()
		f.ready = true
		f.mu.Unlock()
	}()
	return f
}

func (f *Function) Start(_ context.Context, cfg map[string]string) error {
	f.mu.Lock()
	defer f.mu.Unlock()
	f.startCalled = true
	f.configValue = cfg["TEST_CONFIG_VALUE"]
	return nil
}

func (f *Function) Ready(_ context.Context) (bool, error) {
	f.mu.Lock()
	defer f.mu.Unlock()
	return f.ready, nil
}

func (f *Function) Alive(_ context.Context) (bool, error) {
	f.mu.Lock()
	defer f.mu.Unlock()
	return f.alive, nil
}

func (f *Function) Handle(w http.ResponseWriter, r *http.Request) {
	if r.URL.Path == "/set-unhealthy" {
		f.mu.Lock()
		f.alive = false
		f.mu.Unlock()
		fmt.Fprintln(w, "UNHEALTHY")
		return
	}
	f.mu.Lock()
	defer f.mu.Unlock()
	if f.startCalled {
		fmt.Fprintf(w, "START_OK:%s READY=%t ALIVE=%t NONCE=%s", f.configValue, f.ready, f.alive, f.nonce)
	} else {
		fmt.Fprintf(w, "START_NOT_CALLED READY=%t ALIVE=%t NONCE=%s", f.ready, f.alive, f.nonce)
	}
}
